pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,474 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_primitives::{SECONDS_PER_MONTH, SECONDS_PER_YEAR};
+use cfg_traits::{interest::InterestRate, Seconds};
+use chrono::{DateTime, Datelike, Days, Months, NaiveDate, TimeDelta};
+use chronoutil::DateRule;
+use frame_support::pallet_prelude::RuntimeDebug;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{
+		EnsureAddAssign, EnsureDiv, EnsureFixedPointNumber, EnsureInto, EnsureSub, EnsureSubAssign,
+	},
+	ArithmeticError, DispatchError, FixedPointNumber, FixedPointOperand,
+};
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time, in secs
+	Fixed {
+		/// Secs when maturity ends
+		date: Seconds,
+		/// Extension in secs, without special permissions
+		extension: Seconds,
+	},
+}
+
+impl Maturity {
+	pub fn fixed(date: Seconds) -> Self {
+		Self::Fixed { date, extension: 0 }
+	}
+
+	pub fn date(&self) -> Seconds {
+		match self {
+			Maturity::Fixed { date, .. } => *date,
+		}
+	}
+
+	pub fn is_valid(&self, now: Seconds) -> bool {
+		match self {
+			Maturity::Fixed { date, .. } => *date > now,
+		}
+	}
+
+	pub fn extends(&mut self, value: Seconds) -> Result<(), ArithmeticError> {
+		match self {
+			Maturity::Fixed { date, extension } => {
+				date.ensure_add_assign(value)?;
+				extension.ensure_sub_assign(value)
+			}
+		}
+	}
+}
+
+fn seconds_to_days(seconds: Seconds) -> Result<u32, DispatchError> {
+	Ok(TimeDelta::try_seconds(seconds.ensure_into()?)
+		.ok_or(DispatchError::Other("Precission error with seconds"))?
+		.num_days()
+		.ensure_into()?)
+}
+
+fn seconds_to_date(date_in_seconds: Seconds) -> Result<NaiveDate, DispatchError> {
+	Ok(DateTime::from_timestamp(date_in_seconds.ensure_into()?, 0)
+		.ok_or(DispatchError::Other("Invalid date in seconds"))?
+		.date_naive())
+}
+
+fn date_to_seconds(date: NaiveDate) -> Result<Seconds, DispatchError> {
+	Ok(date
+		.and_hms_opt(0, 0, 0)
+		.ok_or(DispatchError::Other("Invalid h/m/s"))?
+		.and_utc()
+		.timestamp()
+		.ensure_into()?)
+}
+
+fn seconds_in_month(date: NaiveDate) -> Result<Seconds, DispatchError> {
+	todo!()
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum ReferenceDate {
+	/// Payments are expected every period relative to a specific date.
+	/// E.g. if the period is monthly and the specific date is Mar 3, the
+	/// first interest payment is expected on Apr 3.
```

Comment:
I just checked back with Dennis and the following would be expected for interest payments:

* Origination date: Anytime during a month
* First interest payment: Last day of month of origination
* Following interest payments: Last day of month
* Last interest payment: Day of maturity

We can neglect only counting business days as valid repayment days for now. That is an edge we can cover by adding soft boundaries to penalties later down the road.

Reply:
> - First interest payment: Last day of month of origination
> - Following interest payments: Last day of month

Yes, but I talked with @hieronx that ideally, it could be any day, if specified.

> - Last interest payment: Day of maturity

Ok, this is different to what I thought 👍🏻